### ISSUE TITLE
Errors when building on windows

### ISSUE THREAD

**Issue by Thalaivar** [kahypar/kahypar#83]
I was following the instructions to build the python interface on windows and after running ` cmake --build . --config Release` in the python folder under `build`, I get the following errors. It looks like some syntax errors

```
(base) PS D:\Downloads\kahypar\build> cmake --build . --config Release
Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule D:/Downloads/kahypar/python/CMakeLists.txt
cl : command line warning D9025: overriding '/W4' with '/w' [D:\Downloads\kahypar\build\python\kahypar_python.vcxproj]
  module.cpp
D:\Downloads\kahypar\kahypar/partition/context.h(458,14): error C2039: 'accumulate': is not a member of 'std' [D:\Downl
oads\kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar/definitions.h(62): message : see declaration of 'std' [D:\Downloads\kahypar\build\python\k
ahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar/partition/context.h(458,1): error C3861: 'accumulate': identifier not found [D:\Downloads\
kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar/partition/context.h(680,12): error C2039: 'accumulate': is not a member of 'std' [D:\Downl
oads\kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar/definitions.h(62): message : see declaration of 'std' [D:\Downloads\kahypar\build\python\k
ahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar/partition/context.h(680,22): error C3861: 'accumulate': identifier not found [D:\Downloads
\kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar\datastructure/hypergraph.h(438,1): error C2672: 'kahypar::ds::end': no matching overloaded
 function found [D:\Downloads\kahypar\build\python\kahypar_python.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include\utility(331): message :
deWeight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::Vertex<kahypar::ds::
GenericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::
PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::HypernodeTraits,kahypar::ds::GenericHypergraph<kahypar::Hyperno
deID,kahypar::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,k
ahypar::meta::Empty>::AdditionalHypernodeData>>' being compiled [D:\Downloads\kahypar\build\python\kahypar_python.vcxpr
oj]
D:\Downloads\kahypar\kahypar/io/hypergraph_io.h(194): message : see reference to class template instantiation 'std::pai
eight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::HypergraphElementIterator<const kahypar::ds::Gen
ericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::Par
titionID,kahypar::meta::Empty,kahypar::meta::Empty>::Vertex<kahypar::ds::GenericHypergraph<kahypar::HypernodeID,kahypar
::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta
::Empty>::HypernodeTraits,kahypar::ds::GenericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWe
ight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::AdditionalHypernodeData>
>,kahypar::ds::GenericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeW
eight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::HypergraphElementIterator<const kahypar::ds::Gen
ericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::Par
titionID,kahypar::meta::Empty,kahypar::meta::Empty>::Vertex<kahypar::ds::GenericHypergraph<kahypar::HypernodeID,kahypar
::HyperedgeID,kahypar::HypernodeWeight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta
::Empty>::HypernodeTraits,kahypar::ds::GenericHypergraph<kahypar::HypernodeID,kahypar::HyperedgeID,kahypar::HypernodeWe
ight,kahypar::HyperedgeWeight,kahypar::PartitionID,kahypar::meta::Empty,kahypar::meta::Empty>::AdditionalHypernodeData>
>>' being compiled [D:\Downloads\kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\kahypar\datastructure/hypergraph.h(441,1): error C2672: 'kahypar::ds::begin': no matching overload
ed function found [D:\Downloads\kahypar\build\python\kahypar_python.vcxproj]
D:\Downloads\kahypar\external_tools\WHFC\util\meta.h(12,10): fatal error C1083: Cannot open include file: 'cxxabi.h': N
o such file or directory [D:\Downloads\kahypar\build\python\kahypar_python.vcxproj]
```

**Comment by SebastianSchlag** [comment]
Hi,
thanks for reaching out.

The error related to `std::accumulate` indeed hinted a missing include, which I fixed in https://github.com/kahypar/kahypar/commit/9ea1e4bcacb5f3e3835916d3c736178cc782297a. 

@larsgottesbueren: The error regarding `cxxabi.h` comes from `WHFC\util\meta.h`. Could you also provide a non-gcc-specific version of `Implementation::type`  that doesn't need the include and doesn't call `abi::__cxa_demangle`?

@Thalaivar: We mainly focus on gcc / clang support and Windows support in general is a bit lacking. Could you use one of those compilers instead of MSVC? Alternatively, you could also try to use the Windows Subsystem for Linux.

**Comment by Roger-luo** [comment]
just FYI this is also blocking the Julia extension support in https://github.com/kahypar/KaHyPar.jl/issues/12, I updated the corresponding JLL package to the official repo instead of the custom fork here: https://github.com/JuliaPackaging/Yggdrasil/pull/3111

PS. I have no idea why the registered binary is a custom fork not the official implementation

**Comment by giordano** [comment]
Other errors when building for Windows with MinGW:
```
[ 98%] Building CXX object lib/CMakeFiles/kahypar.dir/libkahypar.cc.obj
cd /workspace/srcdir/kahypar/build/lib && /opt/bin/x86_64-w64-mingw32-libgfortran5-cxx11/x86_64-w64-mingw32-g++ --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/  -DBOOST_ALL_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK=1 -DKAHYPAR_ENABLE_HEAVY_COARSENING_ASSERTIONS -DKAHYPAR_ENABLE_HEAVY_DATA_STRUCTURE_ASSERTIONS -DKAHYPAR_ENABLE_HEAVY_INITIAL_PARTITIONING_ASSERTIONS -DKAHYPAR_ENABLE_HEAVY_PREPROCESSING_ASSERTIONS -DKAHYPAR_ENABLE_HEAVY_REFINEMENT_ASSERTIONS -Dkahypar_EXPORTS @CMakeFiles/kahypar.dir/includes_CXX.rsp -W -Wall -Wextra  -Wunused  -Wmaybe-uninitialized -Wfatal-errors -Wcast-qual -Woverloaded-virtual -Wredundant-decls -Winit-self -pedantic -DPARANOID  -Weffc++ -Wno-unused-function -mcrc32 -Wno-register -O3 -DNDEBUG -O3   -std=gnu++17 -o CMakeFiles/kahypar.dir/libkahypar.cc.obj -c /workspace/srcdir/kahypar/lib/libkahypar.cc
In file included from /workspace/srcdir/kahypar/lib/libkahypar.cc:21:
/workspace/srcdir/kahypar/lib/../include/libkahypar.h:100:18: warning: redundant redeclaration of ‘void kahypar_hypergraph_free(kahypar_hypergraph_t*)’ in same scope [-Wredundant-decls]
  100 | KAHYPAR_API void kahypar_hypergraph_free(kahypar_hypergraph_t* kahypar_hypergraph);
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/../include/libkahypar.h:63:18: note: previous declaration of ‘void kahypar_hypergraph_free(kahypar_hypergraph_t*)’
   63 | KAHYPAR_API void kahypar_hypergraph_free(kahypar_hypergraph_t* hypergraph);
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:31:20: warning: ‘kahypar_context_t* kahypar_context_new()’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   31 | kahypar_context_t* kahypar_context_new() {
      |                    ^~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:35:6: warning: ‘void kahypar_set_custom_target_block_weights(kahypar_partition_id_t, const kahypar_hypernode_weight_t*, kahypar_context_t*)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   35 | void kahypar_set_custom_target_block_weights(const kahypar_partition_id_t num_blocks,
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:48:6: warning: ‘void kahypar_context_free(kahypar_context_t*)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   48 | void kahypar_context_free(kahypar_context_t* kahypar_context) {
      |      ^~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:55:6: warning: ‘void kahypar_hypergraph_free(kahypar_hypergraph_t*)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   55 | void kahypar_hypergraph_free(kahypar_hypergraph_t* kahypar_hypergraph) {
      |      ^~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:63:6: warning: ‘void kahypar_configure_context_from_file(kahypar_context_t*, const char*)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   63 | void kahypar_configure_context_from_file(kahypar_context_t* kahypar_context,
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:69:6: warning: ‘void kahypar_set_fixed_vertices(kahypar_hypergraph_t*, const kahypar_partition_id_t*)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   69 | void kahypar_set_fixed_vertices(kahypar_hypergraph_t* kahypar_hypergraph,
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:79:23: warning: ‘kahypar_hypergraph_t* kahypar_create_hypergraph_from_file(const char*, kahypar_partition_id_t)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
   79 | kahypar_hypergraph_t* kahypar_create_hypergraph_from_file(const char* file_name, const kahypar_partition_id_t num_blocks) {
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/srcdir/kahypar/lib/libkahypar.cc:92:35: error: function ‘kahypar_hypergraph_t* kahypar_create_hypergraph(kahypar_partition_id_t, kahypar_hypernode_id_t, kahypar_hyperedge_id_t, const size_t*, const kahypar_hyperedge_id_t*, const kahypar_hyperedge_weight_t*, const kahypar_hypernode_weight_t*)’ definition is marked dllimport
   92 | KAHYPAR_API kahypar_hypergraph_t* kahypar_create_hypergraph(const kahypar_partition_id_t num_blocks,
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
make[2]: *** [lib/CMakeFiles/kahypar.dir/build.make:84: lib/CMakeFiles/kahypar.dir/libkahypar.cc.obj] Error 1
```

**Comment by larsgottesbueren** [comment]
> Hi,
> thanks for reaching out.
> 
> The error related to `std::accumulate` indeed hinted a missing include, which I fixed in [9ea1e4b](https://github.com/kahypar/kahypar/commit/9ea1e4bcacb5f3e3835916d3c736178cc782297a).
> 
> @larsgottesbueren: The error regarding `cxxabi.h` comes from `WHFC\util\meta.h`. Could you also provide a non-gcc-specific version of `Implementation::type` that doesn't need the include and doesn't call `abi::__cxa_demangle`? I've done something similar in `template_parameter_to_string.h`.
> 
> @Thalaivar: We mainly focus on gcc / clang support and Windows support in general is a bit lacking. Could you use one of those compilers instead of MSVC? Alternatively, you could also try to use the Windows Subsystem for Linux.

I just removed the demangle call from the whfc library, and updated kahypar to use the new version.
